prow/tide/contextregister.go

```diff
@@ -0,0 +1,145 @@
+func newExpectedContext(c string) Context {
+	return Context{
+		Context:     githubql.String(c),
+		State:       githubql.StatusStateExpected,
+		Description: githubql.String(""),
+	}
+}
+
+// contextRegister implements contextChecker and allow registering of required and optional contexts.
+type contextRegister struct {
+	lock                sync.RWMutex
+	required, optional  sets.String
+	skipUnknownContexts bool
+}
+
+// newContextRegister instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegister(skipUnknownContexts bool) *contextRegister {
+	r := &contextRegister{
+		required:            sets.NewString(),
+		optional:            sets.NewString(),
+		skipUnknownContexts: skipUnknownContexts,
+	}
+	r.registerOptionalContexts(statusContext)
+	return r
+}
+
+// newContextRegisterFromPolicy instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegisterFromPolicy(policy *config.Policy) *contextRegister {
+	r := newContextRegister(false)
+	if policy != nil {
+		if policy.SkipUnknownContexts != nil {
+			r.skipUnknownContexts = *policy.SkipUnknownContexts
+		}
+		if policy.Protect != nil && *policy.Protect {
+			r.registerRequiredContexts(policy.Contexts...)
+		} else if r.skipUnknownContexts {
+			r.registerRequiredContexts(policy.Contexts...)
+		}
+
+	}
+	return r
+}
+
+// newContextRegisterFromPolicy instantiates a new contextRegister and register tide as optional by default
+// and uses Prow Config to find optional and required tests, as well as skipUnknownContexts.
+func newContextRegisterFromConfig(org, repo, branch string, c *config.Config) (*contextRegister, error) {
+	policy, err := c.GetBranchProtection(org, repo, branch)
+	if err != nil {
+		return nil, err
+	}
+	_, optional := config.BranchRequirements(org, repo, branch, c.Presubmits)
+	r := newContextRegisterFromPolicy(policy)
+	r.registerOptionalContexts(optional...)
+	return r, nil
+}
+
+// ignoreContext checks whether a context can be ignored.
+// Will return true if
+// - context is registered as optional
+// - required contexts are registered and the context provided is not required
+// Will return false otherwise. Every context is required.
+func (r *contextRegister) ignoreContext(c Context) bool {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	if r.optional.Has(string(c.Context)) {
+		return true
+	}
+	if r.required.Has(string(c.Context)) {
+		return false
+	}
+	if r.skipUnknownContexts {
+		return true
+	}
+	return false
+}
+
+// missingRequiredContexts discard the optional contexts and only look of extra required contexts that are not provided.
+func (r *contextRegister) missingRequiredContexts(contexts []Context) []Context {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+	if r.required.Len() == 0 {
```

Comment:
Won't the `sets.Difference()` make this happen automatically? No need for this check

Reply:
it will but why bother doing all that when you already know the answer.